File: leetcode-daily/07-January/Maximum Product of Splitted Binary Tree.java
class Solution {

    long ans = 0;
    long mod = (long)1e9 + 7;

    long getTotalSum(TreeNode root) {
        if (root == null) return 0;
        return root.val + getTotalSum(root.left) + getTotalSum(root.right);
    }

    long dfs(TreeNode root, long total) {
        if (root == null) return 0;

        long left = dfs(root.left, total);
        long right = dfs(root.right, total);

        long subTreeSum = left + right + root.val;
        ans = Math.max(ans, subTreeSum * (total - subTreeSum));

        return subTreeSum;
    }

    public int maxProduct(TreeNode root) {
        long total = getTotalSum(root);
        dfs(root, total);
        return (int)(ans % mod);
    }
}

File: leetcode-daily/07-January/Maximum Product of Splitted Binary Tree.cpp
class Solution {
public:
    long long ans = 0;
    const long long mod = 1e9 + 7;

    long long getTotal(TreeNode* root) {
        if (!root) return 0;
        return root->val + getTotal(root->left) + getTotal(root->right);
    }

    long long dfs(TreeNode* root, long long total) {
        if (!root) return 0;

        long long left = dfs(root->left, total);
        long long right = dfs(root->right, total);

        long long subTreeSum = left + right + root->val;
        ans = max(ans, subTreeSum * (total - subTreeSum));

        return subTreeSum;
    }

    int maxProduct(TreeNode* root) {
        long long total = getTotal(root);
        dfs(root, total);
        return ans % mod;
    }
};
